man/image2xy.1
.TH IMAGE2XY "1" "July 2015" "0.56" "astrometry.net"
.SH NAME
image2xy \- Source extractor.
.SH SYNOPSIS
.B image2xy
[options]
\fIfitsname.fits\fR
.SH DESCRIPTION
Read a FITS file, find objects, and write out X, Y, FLUX to fitsname.xy.fits.
.SH OPTIONS
.TP
\fB\-e\fR \fIextension\fR
Read from a single FITS extension
.TP
\fB\-O\fR
Overwrite existing output file.
.TP
\fB\-o\fR \fIfile\fR
Write XYlist to given filename.
.TP
\fB\-L\fR \fIorder\fR
Lanczos order
.TP
\fB\-8\fR
Don't use optimization for byte (u8) images.
.TP
\fB\-H\fR
Downsample by a factor of 2 before running simplexy.
.TP
\fB\-d\fR \fIfactor\fR
Downsample by an integer factor before running simplexy.
.TP
\fB\-D\fR \fIfactor\fR
Downsample, if necessary, by this many factors of two.
.TP
\fB\-s\fR \fIscale\fR\fR
Set median\-filter box size (default 100 pixels)
.TP
\fB\-w\fR \fIwidth\fR
Set Gaussian PSF sigma (default 1 pixel)
.TP
\fB\-g\fR \fIsigma\fR
Set image noise level
.TP
\fB\-p\fR \fIsignificance\fR
Set significance level of peaks in sigmas (default 8 sigmas)
.TP
\fB\-a\fR \fIsaddle\fR
Set "saddle" level joining peaks i sigmas (default 5 sigmas)
.TP
\fB\-P\fR \fIplane\fR
Pull out a single plane of a multi-color image (default: first plane)
.TP
\fB\-b\fR
Don't do (median-based) background subtraction
.TP
\fB\-G\fR \fIbackground\fR
Subtract this 'global' background value; implies \fB\-b\fR
.TP
\fB\-m\fR
Set maximum extended object size for deblending (default 2000 pixels)
.TP
\fB\-S\fR \fIfile\fR
Save background\-subtracted image to this filename (FITS float image)
.TP
\fB\-B\fR \fIfile\fR
Save background image to filename
.TP
\fB\-U\fR \fIfile\fR
Save smoothed background\-subtracted image to filename
.TP
\fB\-M\fR \fIfile\fR
Save mask image to filename
.TP
\fB\-C\fR \fIfile\fR
Save connected\-components image to filename
.TP
\fB\-v\fR
Verbose. repeat for more and more verboseness
.TP
image2xy 'file.fits[1]'
Process first extension.
.TP
image2xy 'file.fits[2]'
Process second extension
.TP
image2xy file.fits+2
Same as above
.SH AUTHOR
The Astrometry.net team. Principal investigators are David W. Hogg (NYU) and
Dustin Lang (CMU).
.SH SEE ALSO
http://astrometry.net


Changelog
0.56:
	implement anwcs_write_to
	Makefile files for install-indexes
	add -std=gnu89 to default CFLAGS
	add clip_wcs() method

0.55:
	Fix bug in WCS numpy array iterators

0.54:
	Try importing pyfits and then astropy.io.fits

0.53:
	Update wcs-pv2sip and include in default build
	Fixes for git version strings
	Read and write linear SIP distortion terms
	multiindex: don't assert when reloading already-loaded file

0.52:
	Fix release version strings in released versions.

0.51:
	Fix half-pixel bug in "solve-field --crpix-center": the center was
	set to 1 + {W,H}/2, not 0.5 + {W,H}/2.  (The WCS was still
	correct, the CRPIX was set to a half pixel in each dimension away
	from the center pixel.




blind/image2xy.py
# This file is part of the Astrometry.net suite.
# Copyright 2007 Keir Mierle and David W. Hogg.
#
# The Astrometry.net suite is free software; you can redistribute
# it and/or modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2, or
# (at your option) any later version.
#
# The Astrometry.net suite is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with the Astrometry.net suite ; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA

"""
NAME:
      image2xy

PURPOSE:
      Extract sources from a FITS file

INPUTS:
      Takes a single FITS file as input

OPTIONAL INPUTS:

KEYWORD PARAMETERS:

OUTPUTS:

OPTIONAL OUTPUTS:

COMMENTS:

MODIFICATION HISTORY:
       K. Mierle, 2007-Jan - Initial version based on image2xy.c
       Hogg, 2007-May - simplexy options change
"""

# You need ctypes and a recent (1.0) numpy for this to work. I've included
# pyfits so you don't have to. 

try:
    import pyfits
except ImportError:
    try:
        from astropy.io import fits as pyfits
    except ImportError:
        raise ImportError("Cannot import either pyfits or astropy.io.fits")
import sys
import scipy
import os
import time
from simplexy import simplexy

# Default settings
dpsf = 1.0
plim = 8.0
dlim = dpsf
saddle = 5.0
maxper = 1000
maxnpeaks = 10000
maxsize = 1000
halfbox= 100

def source_extract(image_data, srcext=None):


    x,y,flux,sigma = simplexy(image_data, dpsf=dpsf, plim=plim,
                              dlim=dlim, saddle=saddle, maxper=maxper,
                              maxnpeaks=maxnpeaks, maxsize=maxsize,
                              halfbox=halfbox)

    print 'simplexy: shapes',x.shape, y.shape

    cx = pyfits.Column(name='X', format='E', array=x, unit='pix')
    cy = pyfits.Column(name='Y', format='E', array=y, unit='pix')
    cflux = pyfits.Column(name='FLUX', format='E', array=flux)

    tbhdu = pyfits.new_table([cx, cy, cflux])
    h = tbhdu.header
    h.add_comment('Parameters used in source extraction')
    h.update('dpsf',     dpsf, 'Gaussian psf width')
    h.update('plim',     plim, 'Significance to keep')
    h.update('dlim',     dlim, 'Closest two peaks can be')
    h.update('saddle',   saddle, 'Saddle in difference (in sig)')
    h.update('maxper',   maxper, 'Max num of peaks per object')
    h.update('maxpeaks', maxnpeaks, 'Max num of peaks total')
    h.update('maxsize',  maxsize, 'Max size of extended objects')
    h.update('halfbox',  halfbox, 'Half-size of sliding sky window')
    if srcext != None:
        h.update('srcext', i, 'Extension number in src image')
    h.update('estsigma', sigma, 'Estimated source image variance')
    h.add_comment('The X and Y points are specified assuming 1,1 is ')
    h.add_comment('the center of the leftmost bottom pixel of the ')
    h.add_comment('image in accordance with the FITS standard.')
    h.add_comment('Extracted by image2xy.py')
    h.add_comment('on %s %s' % (time.ctime(), time.tzname[0]))
    cards = tbhdu.header.ascardlist()
    cards['TTYPE1'].comment = 'X coordinate'
    cards['TTYPE2'].comment = 'Y coordinate'
    cards['TTYPE3'].comment = 'Flux of source'

    return tbhdu

def extract(fitsfile):
    outfile = pyfits.HDUList()

    # Make empty HDU; no image
    outfile.append(pyfits.PrimaryHDU()) 

    for i, hdu in enumerate(fitsfile):
        if (i == 0 and hdu.data != None) or isinstance(hdu, pyfits.ImageHDU):
            print hdu.data.shape+(i,)
            if i == 0:
                print 'Image: Primary HDU (number 0) %sx%s' % hdu.data.shape
            else:
                print 'Image: Extension HDU (number %s) %sx%s' % tuple((i,)+hdu.data.shape)

            tbhdu = source_extract(image_data)

            outfile.append(tbhdu)

    return x,y,flux,sigma, outfile

if __name__ == '__main__':
    if len(sys.argv) != 2:
        print "Usage: image2xy.py image.fits"
    else:
        infile = sys.argv[1]
        fitsfile = pyfits.open(infile)
        x,y,flux,sigma, outfile = extract(fitsfile)
        newfile = infile.replace('.fits','.xy.fits')
        try:
            outfile.writeto(newfile)
        except IOError:
            # File probably exists
            print 'File %s appears to already exist; deleting!' % newfile
            import os
            os.unlink(newfile)
            outfile.writeto(newfile)
